supervisor/api/snapshots.py

```diff
@@ -180,25 +180,39 @@ async def download(self, request):
     @api_process
     async def upload(self, request):
         """Upload a snapshot file."""
-        with TemporaryDirectory(dir=str(self.sys_config.path_tmp)) as temp_dir:
-            tar_file = Path(temp_dir, "snapshot.tar")
-
-            try:
-                with tar_file.open("wb") as snapshot:
-                    async for data in request.content.iter_any():
-                        snapshot.write(data)
+        chunks = []
+        temp_dir = TemporaryDirectory(dir=str(self.sys_config.path_tmp))
+
+        def _write_chunks():
+            """Write chunks to tempfile and return the filepath."""
+            tar_file = Path(temp_dir.name, "snapshot.tar")
+            with tar_file.open("wb") as snapshot:
+                for chunk in chunks:
+                    snapshot.write(chunk)
+            return tar_file
+
+        reader = await request.multipart()
+        contents = await reader.next()
+        try:
+            while True:
+                chunk = await contents.read_chunk()
+                if not chunk:
+                    break
+                chunks.append(chunk)
```

Comment:
Don't cache it on the memory, write it directly to the file as a stream

Reply:
https://github.com/home-assistant/supervisor/pull/2076#discussion_r493724676
If we do that we are doing I/O in the eventloop

Reply:
Supervisor handle all file request in I/O. File handling is a bit bit complexer and not every file access doing also IO to disk, just the close of the file doing IO.

Reply:
The supervisor have no issue with micro blocks, unlike HA

Reply:
Maybe we can switch to an good library with low level file handling for asyncio loop (and not just execute stuff on a executor)

Reply:
Aiit :+1: 
I'll revert this commit then :)
<[`7f26b43` (#2076)](https://github.com/home-assistant/supervisor/pull/2076/commits/7f26b432cb057d4fb08e11993edf5f3a008e7c5e)>